### DISCUSSION TITLE
Define the model of the objects that a data bucket stores in an array

### DISCUSSION THREAD

**Discussion by LukMas** [mockoon/mockoon#1427] [Ideas]
Would it be possible to add a feature that allows to define the model of the data stored in a data bucket, when an array of objects is used, to allow a better customization of where the id of the object is set? 

Actually we are mocking our APIs with Mockoon, but we have a problem with some CRUD endpoints. We are using the JSON:API specification to exchange our data. When a new resource is created (with POST) we MUST send the following content: 
```
{
  "data": {
    "type": "photos",
    "attributes": {
      "title": "Ember Hamster",
      "src": "http://example.com/images/productivity.png"
    },
    "relationships": {
      "photographer": {
        "data": { "type": "people", "id": "9" }
      }
    }
  }
}
```
The server MUST return the following response: 
```
{
  "data": {
    "type": "photos",
    "id": "550e8400-e29b-41d4-a716-446655440000",
    "attributes": {
      "title": "Ember Hamster",
      "src": "http://example.com/images/productivity.png"
    },
    "links": {
      "self": "http://example.com/photos/550e8400-e29b-41d4-a716-446655440000"
    }
  }
}
```
The problem is that Mockoon manages the data in a different way, returning the following response: 
```
{
  "id": "550e8400-e29b-41d4-a716-446655440000",
  "data": {
    "type": "photos",
    "attributes": {
      "title": "Ember Hamster",
      "src": "http://example.com/images/productivity.png"
    },
    "links": {
      "self": "http://example.com/photos/550e8400-e29b-41d4-a716-446655440000"
    }
  }
}
```
I've already adapted the content of the data bucket to support how Mockoon handles the ids, and it works for GETs because I could manipulate the content before returning it (with a custom endpoint). 

The problem is that there's no way to make it working for PATCHs and PUTs (actually also the POST, if the content of the response is reused, is wrong), because Mockoon doesn't understand what the client sends. I also can't customized the endpoint because I would completely lose the manipulation of the content of the data bucket. 

I don't know if I miss some knowledge to handle correctly this situation. But, I don't think there is a way to manipulate the request before storing the data in the data bucket. That means, POST, PATCH and PUT could not be manipulated in a way that the data reflects the structure we require to use.

Hence, I would like to know if it's possible to add a new feature that allows to set, for complex object, a model of the data stored in the data bucket. This would the enable also the possibility to set a custom id that the data bucket manages, opening for a better support for APIs with complex object.

**Comment by 255kb** [comment]
The CRUD/bucket feature was thought as a way to get a basic REST/CRUD API up and running. We mostly got inspired by json-server, which offers a bit more features like relationships between data. 
Some users suggested improvements like allowing ["wrapping" of the returned data](https://github.com/mockoon/mockoon/issues/1419), and it will probable evolve, I'm not sure the feature can support each and every specification out there.

**Reply by LukMas** [reply]
I understand and I think that, once it's possible to post-process the response at least the problem with the POST would be solved. 
I send my data and Mockoon stores it like this: 
```
{
  "id": "550e8400-e29b-41d4-a716-446655440000",
  "data": {
    "type": "photos",
    "attributes": {
      "title": "Ember Hamster",
      "src": "http://example.com/images/productivity.png"
    },
    "links": {
      "self": "http://example.com/photos/550e8400-e29b-41d4-a716-446655440000"
    }
  }
}
```
and the post-process manipulates it in a way that the `id` goes back into the `data` object. 

The problem remains anyway for PUT/PATH because what I send has the `Id` inside `data`, while Mockoon expects it outside. Hence there's no way to simulate this actions because the object would never match. 

A schema, or just a way to define the `id` as a nested property would solve the problem. 
I don't know, maybe a special construct like this for nested properties. 
```
Id (a simple id)
data->id (when id is nested within the data object)
```
It would be really simple to document, but it would allow to have Id that are nested inside the objects that are stored.

**Reply by LukMas** [reply]
By the way, I even could try to help for the project for this (or other features), given that I extensively use your software at work.

**Reply by 255kb** [reply]
I'm wondering, you can already provide the "crudKey", the id property on which the operations are based: 
![image](https://github.com/mockoon/mockoon/assets/7489814/12b65f63-d46c-425c-b465-ad5558d44503)

For now, this option only supports properties at the root level of objects. If this would support the dot notation that is supported elsewhere (in templating helpers for example) it seems it would solve part of your issue no? 
![image](https://github.com/mockoon/mockoon/assets/7489814/a18155d4-ca17-43b4-b01d-0f995659736b)

The amount of changes required wouldn't be too big I think. 

> By the way, I even could try to help for the project for this (or other features), given that I extensively use your software at work.

You are welcome to contribute, it's really appreciated :)

**Reply by LukMas** [reply]
Yes, this would solve the problem. 
I was thinking about this in the last days. I don't really need a model of the data, just the way to say where is the _crudKey_ is enough. Then I could store the data in the way that I expect to have it.

Actually the response would require anyway some work to support `relationships` and `includes`, but I could take directly the content of the data bucket and add information to it (while now I've to bend it to be compliant to JSON:API).

**Reply by 255kb** [reply]
I opened an issue to track this https://github.com/mockoon/mockoon/issues/1433

**Reply by LukMas** [reply]
Super, thank you!

**Reply by LukMas** [reply]
Thank you very much for this!
Sorry, I would have helped to implement, but actually I'm too much busy to do something.

**Reply by 255kb** [reply]
no problem, this is quite a "small" change. Let's see if it solves this use case or if it needs more fine tuning :)
